libraries/common/map.effekt

```diff
@@ -0,0 +1,718 @@
+// O(log N)
+def getIndex[K, V](m: Map[K, V], n: Int): Option[(K, V)] = {
+  m match {
+    case Tip() => None()
+    case Bin(size, k, v, l, r) =>
+      val sizeL = l.size()
+      genericCompare(sizeL, n) match {
+        case Less() => r.getIndex(n - (sizeL + 1))
+        case Greater() => l.getIndex(n)
+        case Equal() => Some((k, v))
+      }
+  }
+}
+
+// Construct a new map which contains all elements of `m1`
+// except those where the key is found in `m2`.
+//
+// O(???)
+def difference[K, V](m1: Map[K, V], m2: Map[K, V]): Map[K, V] = {
+  (m1, m2) match {
+    case (Tip(), m2) => Tip()
+    case (m1, Tip()) => m1
+    case (m1, Bin(_, k, _, l2, r2)) =>
+      val (l1, r1) = m1.split(k)
+      val leftDiff = l1.difference(l2)
+      val rightDiff = r1.difference(r2)
+      if ((leftDiff.size() + rightDiff.size()) == m1.size()) { m1 }
+      else { link2(leftDiff, rightDiff) }
+  }
+}
+
+// Construct a new map which contains the elements of both `m1` and `m2`.
+// When a key is associated with a value in both `m1` and `m2`, the new value is determined using the `combine` function.
+//
+// O(???)
+def union[K, V](m1: Map[K, V], m2: Map[K, V]) { combine : (K, V, V) => V }: Map[K, V] = {
+  // Internal function similar to `putWithKey`, but right-biased.
+  // Only used here, recursively.
+  def putWithKeyR(m: Map[K, V], k: K, v: V): Map[K, V] = {
+    m match {
+      case Tip() => singleton(k, v)
+      case Bin(size, k2, v2, l, r) =>
+        genericCompare(k, k2) match {
+          case Less() => balance(k2, v2, l.putWithKeyR(k, v), r)
+          case Greater() => balance(k2, v2, l, r.putWithKeyR(k, v))
+          case Equal() => Bin(size, k, combine(k2, v2, v), l, r)
+        }
+    }
+  }
+
+  (m1, m2) match {
+    case (_, Tip()) => m1
+    case (_, Bin(_, k, v, Tip(), Tip())) => m1.putWithKeyR(k, v)
+    case (Bin(_, k, v, Tip(), Tip()), _) => m2.putWithKey(k, v){combine}
+    case (Tip(), _) => m2
+    case (Bin(_, k1, v1, l1, r1), _) =>
+      val (l2, optMid, r2) = m2.splitLookup(k1)
+      val leftUnion = union(l1, l2){combine}
+      val rightUnion = union(r1, r2){combine}
+      optMid match {
+        case None() => link(k1, v1, leftUnion, rightUnion)
+        case Some(v2) => link(k1, combine(k1, v1, v2), leftUnion, rightUnion)
+      }
+  }
+}
+
+// Construct a new map which contains the elements of both `m1` and `m2`.
+// When a key is associated with a value in both `m1` and `m2`, the new value is determined using the `combine` function.
+//
+// O(???)
+def union[K, V](m1: Map[K, V], m2: Map[K, V]) { combine : (V, V) => V }: Map[K, V] = {
+  union(m1, m2) { (k, v1, v2) => combine(v1, v2) }
+}
+
+// Construct a new map which contains the elements of both `m1` and `m2`.
+// Left-biased: Uses values from `m1` if there are duplicate keys.
+//
+// O(???)
+def union[K, V](m1: Map[K, V], m2: Map[K, V]): Map[K, V] = {
+  union(m1, m2) { (k, v1, v2) => v1 }
+}
+
+// -------------
+// Internal
+
+val ratio = 2
+val delta = 3
```

Comment:
Missing comparison primitives notwithstanding, this ^ would also require #496 in order to compile for LLVM.